### ISSUE TITLE
pid_exists(0) return False on Linux

### ISSUE THREAD

**Issue by Gnouc** [giampaolo/psutil#844]
On Linux, where `psutil.POSIX` is True, `pid_exists(0)` return False:

```
>>> import psutil
>>> psutil.pid_exists(0)
False
```

In contrast, on FreeBSD, `pid_exists(0)` return True.

Furthermore, PID 0 is not in `psutils.pids()` on Linux, while it does in FreeBSD, that causes the check fail because of [this line](https://github.com/giampaolo/psutil/blob/master/psutil/__init__.py#L1296):

```
def pid_exists(pid):
    """Return True if given PID exists in the current process list.
    This is faster than doing "pid in psutil.pids()" and
    should be preferred.
    """
    if pid < 0:
        return False
    elif pid == 0 and POSIX:
        # On POSIX we use os.kill() to determine PID existence.
        # According to "man 2 kill" PID 0 has a special meaning
        # though: it refers to <<every process in the process
        # group of the calling process>> and that is not we want
        # to do here.
        return pid in pids()
    else:
        return _psplatform.pid_exists(pid)
```

From user perspective, it seems to be weird, on different POSIX systems, the result of `pid_exists(0)` should be consistent, shouldn't it?

IMHO, just remove the second condition check and falling to `_psplatform.pid_exists(pid)` is enough.

**Comment by giampaolo** [comment]
`pid_exists(0)` on Linux returns False simply because PID 0 on Linux does not exist, and that really is it. =) 
Some platforms such as FreeBSD and Windows have PID 0, others such as Linux doesn't, and `pid_exists` reflects this. If you do "ps aux" on Linux you'll notice how PID 0 is not listed.

**Comment by Gnouc** [comment]
@giampaolo If you did
```
$ ps -eaf
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 22:29 ?        00:00:00 /sbin/init splash
cuonglm   3945  3709  0 23:03 pts/1    00:00:00 ps -af -p 1
```

You see parent process ID of init is 0. Also:
```
$ kill -0 0
$ echo "$?"
0
```

That's clearly said that PID 0 is existed, according to `man 2 kill`

**Comment by giampaolo** [comment]
Problem is `/proc/0` directory does not exist so we cannot get any info about it, nor instantiate `psutil.Process(0)`. `psutil.Process(1).ppid()` correctly returns `0` thought. 
Unfortunately it's an inconsistency inherited by the Linux kernel / /proc fs.

On the other hand, the fact that `kill -0 0` does not fail makes me think that `pid_exists` should return `True` though. But then again, that would be inconsistent with what returned by `psutil.pids()`, which does not include any `0` pid.

Mmhh...

**Comment by mrjefftang** [comment]
Pid 0 on Linux does not exist.

Furthermore, if you send a kill with pid=0, you are sending the signal to the entire process group.

From kill(2):
> If pid equals 0, then sig is sent to every process in the process group of the calling process.

From kill(1):
>Specify the list of processes that kill should signal. Each pid can be one of five things:
>
>n
>where n is larger than 0. The process with pid n will be signaled.
>
>0
>
>All processes in the current process group are signaled.
>
>-1
>
>All processes with pid larger than 1 will be signaled.
>
>-n
>
>where n is larger than 1. All processes in process group n are signaled. When an argument of the form '->n' is given, and it is meant to denote a process group, either the signal must be specified first, or the >argument must be preceded by a '--' option, otherwise it will be taken as the signal to send.
>
> commandname

**Comment by Gnouc** [comment]
@mrjefftang It does, it just was not exposed to user space, see http://unix.stackexchange.com/questions/83322/which-process-has-pid-0

**Comment by giampaolo** [comment]
Getting back to this. Since `psutil.Process(0)` fails with NoSuchProcess and  `psutil.pids()` does not return any `0` PID I don't see any practical advantage in making `pid_exists(0)` return `True`.
It would be more "correct", but in the practical sense it would be inconsistent with the rest of the available APIs. In summary, I would say this is one of those case where practicality beats purity.
Closing this out.

**Comment by giampaolo** [comment]
To further elaborate on this, some time ago when I worked on implementing `parents()` method I noticed that regardless on whether the platforms has PID 0 or not, `ppid()` can still return 0. On Linux:
```python
>>> psutil.Process(1).ppid()
0
```
...so basically this means that kernel uses `0` as a value to express "there's no parent pid". What one may argue to return instead is `None` but again, practicality beats purity.
